### PR TITLE
Big chungus speedup

### PR DESCRIPTION
**Pull request by ciebson123**
# Description
Updated chunking so it works faster and is compatible with DDP

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [x] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Comment by ciebson123** [comment]
Multigpu: https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2502&type=run
Speed comparison: https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2468&dash=charts&type=run&compare=EwFgbGA0xA

File: research/conditional/utils/model_utils.py [hunk 1]
```diff
@@ -49,90 +48,96 @@
...
     batch: LLMBatch,
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    scaler: torch.cuda.amp.GradScaler,
     n_chungs: int,
+    backward_pass: bool,
+    num_accumulated_batches: int,
 ):
+    if isinstance(model, DDP):
```
Comment (ludziej) [inline research/conditional/utils/model_utils.py:73]: here we will need to make if work with FSDP as well @kuba-krj FYI (no changes required right now)

**Comment by ludziej** [comment]
> Multigpu: https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2502&type=run Speed comparison: https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2468&dash=charts&type=run&compare=EwFgbGA0xA

It's counter-intuitive that new chungus actually uses more memory? It should be the other way around, and we need chungus specifically for that reason.
Can you also provide multigpu speed comparison? Just run baseline with ddp as well.

**Review by ludziej — CHANGES_REQUESTED** [review]
Code looks nice style-wise, but we need to work this out until it uses less memory. If it does not inherently, I think we should keep the previous chungus as well, but I dont see the reason for that right now.

**Comment by kuba-krj** [comment]
> > Multigpu: https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2502&type=run Speed comparison: https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9a15aa36-1063-49c4-807d-53393abb118a&detailsTab=metadata&shortId=LLMRANDOM-2468&dash=charts&type=run&compare=EwFgbGA0xA
> 
> It's counter-intuitive that new chungus actually uses more memory? It should be the other way around, and we need chungus specifically for that reason.
> Can you also provide multigpu speed comparison? Just run baseline with ddp as well.

I believe here the reason might be that there was another job running on the same card

**Review by Simontwice — CHANGES_REQUESTED** [review]
This is a massive PR, I am assigning myself as second reviewer

File: research/conditional/utils/conditional_trainer.py [hunk 2]
```diff
@@ -210,22 +210,16 @@ def _decode_samples(self, step):
                 iteration=step,
             )
 
-    def _optimize(self, loss, should_apply_gradient=False):
-        # since we sum gradients averaged over multiple smaller batches, we need to normalize here
-        loss /= self.gradient_accumulation_steps
-        if self.gradient_accumulation_steps == 1:
-            self.optimizer.zero_grad()
-        # clear computation graph, store gradients
-        self.scaler.scale(loss).backward()
```
Comment (Simontwice) [inline research/conditional/utils/conditional_trainer.py:219]: what happened here? how is gradient accumulation implemented now?

Comment (ciebson123) [reply research/conditional/utils/conditional_trainer.py:219]: It zeros out gradients after applying them, about 10 lines below. If previous logic was required (zeroing before if no accumulation and after if accumulation) i can try to revert it back

**Comment by Simontwice** [comment]
All right, this calls for a battery of tests. I need links to neptune for every experiment described below (1 before PR, 1 after PR):

1. DDP
2. model parallel
3. gradient accumulation
4. normal checkpointing
5. DDP + gradient accumulation
6. model parallel + gradient accumulation
7. model parallel + normal checkpointing

**Review by Simontwice — CHANGES_REQUESTED** [review]
as described in a comment above

File: research/conditional/utils/conditional_trainer.py (hunk 2 above)
Comment (Simontwice) [reply research/conditional/utils/conditional_trainer.py:222]: As discussed in person, this is acceptable, even better than before.

File: research/conditional/utils/model_utils.py (hunk 1 above)
Comment (Simontwice) [reply research/conditional/utils/model_utils.py:73]: yeah, this PR is ready and FSDP is not, so it should be implemented in the FSDP PR.

File: research/conditional/utils/model_utils.py [hunk 3]
```diff
@@ -49,90 +48,96 @@
...
-                partial_correct_tokens,
-                partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
-            )
-            num_tokens += partial_loss_output.shape[0]
-            total_loss += partial_loss_output.sum()
+            output = model.head(chunged_input)
+            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
+                chunged_gt = chunged_gt.to(output.device)
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:98]: this communication should be moved before the loop. Send the whole `gt` at once.

File: research/conditional/utils/model_utils.py [hunk 4]
```diff
@@ -49,90 +48,96 @@
...
-            (
-                partial_loss_output,
-                partial_correct_tokens,
-                partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
-            )
-            num_tokens += partial_loss_output.shape[0]
-            total_loss += partial_loss_output.sum()
+            output = model.head(chunged_input)
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:96]: rename it to `partial_output`

File: research/conditional/utils/model_utils.py [hunk 5]
```diff
@@ -49,90 +48,96 @@
...
+            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
+                chunged_gt = chunged_gt.to(output.device)
+                partial_loss_full = F.cross_entropy(
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:99]: rename to `partial_loss_unmasked` - "full" is misleading

File: research/conditional/utils/model_utils.py [hunk 6]
```diff
@@ -49,90 +48,96 @@
...
 
-def chungized_llm_loss(
+def chungized_llm_loss_and_backward_pass(
     batch: LLMBatch,
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    scaler: torch.cuda.amp.GradScaler,
     n_chungs: int,
+    backward_pass: bool,
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:70]: from what I've seen in the code, this is not needed. Remove this argument and replace it with `model.train()` and `model.eval()`

File: research/conditional/utils/model_utils.py [hunk 7]
```diff
@@ -49,90 +48,96 @@
...
+    num_accumulated_batches: int,
 ):
+    if isinstance(model, DDP):
+        model = model.module
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:74]: what happens here?

File: research/conditional/utils/model_utils.py [hunk 8]
```diff
@@ -49,90 +48,96 @@
...
 
     with torch.autocast(
         device_type="cuda", enabled=mixed_precision, dtype=torch.float16
     ):
-        encoder_output = model.encoder(input_tokens)
-        chunged_inputs = torch.chunk(encoder_output, n_chungs, dim=0)
+        encoder_output: torch.Tensor = model.encoder(input_tokens)
+        encoder_output_det = encoder_output.detach()
+        if backward_pass:
+            encoder_output_det.requires_grad = True
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:86]: this detachment is due to the fact that you calculate loss manually, right? We couldn't leave the output attached?

File: research/conditional/utils/model_utils.py [hunk 9]
```diff
@@ -49,90 +48,96 @@
...
+                    -1
+                ) * chunged_mask.reshape(-1)
+                partial_correct_tokens = partial_correct_tokens.sum()
+
+            if backward_pass:
+                loss = partial_loss.sum() / num_masked_tokens / num_accumulated_batches
+                with torch.autocast(
+                    device_type="cuda", enabled=False, dtype=torch.float16
+                ):
+                    scaler.scale(loss).backward()
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:117]: isn't this eqivalent to just deleting this context manager? since it's hardcoded to False

File: research/conditional/utils/model_utils.py (hunk 7 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:74]: DDP wraps the torch module and doesn't expose it's parts, so we have to unwrap it to evaluate encoder and head separately

File: research/conditional/utils/model_utils.py [hunk 10]
```diff
@@ -49,90 +48,96 @@
...
-                partial_loss_output,
-                partial_correct_tokens,
-                partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
-            )
-            num_tokens += partial_loss_output.shape[0]
-            total_loss += partial_loss_output.sum()
+            output = model.head(chunged_input)
+            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
```
Comment (Simontwice) [inline research/conditional/utils/model_utils.py:97]: isn't this eqivalent to just deleting this context manager? since it's hardcoded to False

File: research/conditional/utils/model_utils.py (hunk 9 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:117]: This code is wrapped in another context manager that may have it set to true, and here we have to turn it off as per torch standard

File: research/conditional/utils/model_utils.py (hunk 8 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:86]: Yes, if we leave it attached the gradient would pass further into the network and we don't want to do that for performance reasons

File: research/conditional/utils/model_utils.py (hunk 10 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:97]: Same as above, this part needs to have it turned off because of some autocast fuckery which slows it down if it is turned on

File: research/conditional/utils/model_utils.py (hunk 8 above)
Comment (Simontwice) [reply research/conditional/utils/model_utils.py:86]: I see, thanks!

File: research/conditional/utils/model_utils.py (hunk 10 above)
Comment (Simontwice) [reply research/conditional/utils/model_utils.py:97]: noted!

File: research/conditional/utils/model_utils.py (hunk 4 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:96]: Done

File: research/conditional/utils/model_utils.py (hunk 3 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:98]: We don't have this output before loop. Is it true (in model_parallel) that encoder_output and head output will be on the same device?

File: research/conditional/utils/model_utils.py (hunk 5 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:99]: Done

File: research/conditional/utils/model_utils.py (hunk 6 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py]: Done here, probably needs some cleanup in code calling this @Simontwice

**Comment by ciebson123** [comment]
New, comprehensible experiments: https://app.neptune.ai/o/pmtest/org/llm-random/runs/table?viewId=9a94ac8c-b282-48fd-8a2b-5b07d199d9e6&detailsTab=monitoring&shortId=LLMRANDOM-2871&dash=charts&type=run. I know some exps are short, but I needed athena to let me make them.

**Comment by ciebson123** [comment]
Summon the review gods: @ludziej @Simontwice

**Review by Simontwice — APPROVED** [review]
blind accept to remove "Request changes"

**Review by ludziej — APPROVED** [review]
remove request changes

**Comment by ciebson123** [comment]
Screenshot from 2023-12-05 15-32-26
Screenshot from 2023-12-05 15-32-36
Screenshot from 2023-12-05 15-32-44
Screenshot from 2023-12-05 15-32-54
Memory usage comparison:

1. New chungized loss
2. New non chungized loss
3. Old chungized loss
4. Old non chugized loss

Everything looks good

**Comment by ciebson123** [comment]
https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9ab6871a-1bb1-4b8f-a0c9-e3688f4af0cf&dash=charts&compare=CwRgnAzANCQ&chartFilter=ste

Comparison of speed chungus fsdp - no fsdp

**Comment by ludziej** [comment]
LGTM, this PR was here for too long
